File: js/face-remove.js

/**
 * General purpose functions
 */


/**
 * Multiply two matrices
 * @param {Array 2d} a the pre multiplicative matrix
 * @param {Array 2d} b the post multiplicative matrix
 * @returns {Array 2d} result matrix
 */
function matrixMultiply(matrix_1, matrix_2) {
    let result = new Array(matrix_1.length).fill(0);
    for (let i = 0; i < result.length; i++) {
        result[i] = new Array(matrix_2[i].length).fill(0);
        for (let j = 0; j < result[i].length; j++) {
            for (let k = 0; k < matrix_1[i].length; k ++) {
                result[i][j] += matrix_1[i][k] * matrix_2[k][j];
            }
        }
    }
    return result;
}


/**
 * Rotate a 3D vector
 * @param {Array 1d} vector the 3D vector
 * @param {Array 1d} r the 3D rotation
 * @returns {Array 1d} result 3D vector
 */
function rotateVector3D(vector, r) {
    let result = [[vector[0]], [vector[1]], [vector[2]]];
    let rotationMatrixX = [
        [1, 0, 0],
        [0, Math.cos(r[0]), -Math.sin(r[0])],
        [0, Math.sin(r[0]), Math.cos(r[0])]
    ];
    let rotationMatrixY = [
        [Math.cos(r[1]), 0, Math.sin(r[1])],
        [0, 1, 0],
        [-Math.sin(r[1]), 0, Math.cos(r[1])]
    ];
    let rotationMatrixZ = [
        [Math.cos(r[2]), -Math.sin(r[2]), 0],
        [Math.sin(r[2]), Math.cos(r[2]), 0],
        [0, 0, 1]
    ];

    let temp_1 = matrixMultiply(rotationMatrixX, rotationMatrixY);
    temp_1 = matrixMultiply(temp_1, rotationMatrixZ);
    temp_1 = matrixMultiply(temp_1, result);
    return result = [temp_1[0][0], temp_1[1][0], temp_1[2][0]];
}


/**
 * Rotate a 3D vector (inverse)
 * @param {Array 1d} vector the 3D vector
 * @param {Array 1d} r the 3D rotation
 * @returns {Array 1d} result 3D vector
 */
function rotateVector3DInverse(vector, r) {
    let result = [[vector[0]], [vector[1]], [vector[2]]];
    let rotationMatrixX = [
        [1, 0, 0],
        [0, Math.cos(r[0]), Math.sin(r[0])],
        [0, -Math.sin(r[0]), Math.cos(r[0])]
    ];
    let rotationMatrixY = [
        [Math.cos(r[1]), 0, -Math.sin(r[1])],
        [0, 1, 0],
        [Math.sin(r[1]), 0, Math.cos(r[1])]
    ];
    let rotationMatrixZ = [
        [Math.cos(r[2]), Math.sin(r[2]), 0],
        [-Math.sin(r[2]), Math.cos(r[2]), 0],
        [0, 0, 1]
    ];

    let temp_1 = matrixMultiply(rotationMatrixZ, rotationMatrixY);
    temp_1 = matrixMultiply(temp_1, rotationMatrixX);
    temp_1 = matrixMultiply(temp_1, result);
    return result = [temp_1[0][0], temp_1[1][0], temp_1[2][0]];
}


/**
 * Scale a vector
 * @param {Array 1d} vector the nD vector
 * @param {Float} scale
 * @returns {Array 1d} result nD vector
 */
function scaleVector(vector, scale) {
    let result = new Array(vector.length);
    for (let i = 0; i < vector.length; i++) {
        result[i] = vector[i] * scale;
    }
    return result;
}


/**
 * Add two vectors
 * @param {Array 1d} vector_1 an nD vector
 * @param {Array 1d} vector_2 an nD vector
 * @returns {Array 1d} result nD vector
 */
function addVectors(vector_1, vector_2) {
    let result = new Array(vector_1.length);
    for (let i = 0; i < vector_1.length; i++) {
        result[i] = vector_1[i] + vector_2[i];
    }
    return result;
}


/**
 * Get vectors parallel to the sides of a rotated cube
 * @param {Array 1d} s the size of cube
 * @param {Array 1d} r the rotation of cube
 * @returns {Array 3d of Array 1d} result vectors
 */
function getSideVectors(s, r) {
    let temp_1 = rotateVector3D([s[0],0,0], r);
    let temp_2 = rotateVector3D([0,s[1],0], r);
    let temp_3 = rotateVector3D([0,0,s[2]], r);
    return [temp_1, temp_2, temp_3];
}


/**
 * Get an array of points for face
 * @param {Array 1d} p the position of cube
 * @param {Array 1d} s the size of cube
 * @param {Array 1d} r the rotation of cube
 * @param {Integer} face the face number of cube, between 0 and 5 both inclusive
 * @param {Integer} divisions the number of columns and rows you want per face grid
 * @returns {Array of Array 3d} result
 */
function getFacePoints(p, s, r, face, divisions) {
    let side_vectors = getSideVectors(s, r);

    let vector_1 = side_vectors[face % 3];
    let vector_2 = side_vectors[(face + 1) % 3];
    let vector_3 = side_vectors[(face + 2) % 3];
    let sign = (face % 2) ? -1 : 1;

    let offset = addVectors(scaleVector(vector_1, -0.5), scaleVector(vector_2, -0.5));
    offset = addVectors(offset, scaleVector(vector_3, 0.5 * sign));
    offset = addVectors(offset, p);
    let vector_1_div = scaleVector(vector_1, 1 /divisions);
    let vector_2_div = scaleVector(vector_2, 1 / divisions);

    let points = new Array();
    for (let i = 0; i <= divisions; i ++) {
        for (let j = 0; j <= divisions; j ++) {
            let temp_vector = addVectors(scaleVector(vector_1_div, i), scaleVector(vector_2_div, j));
            temp_vector = addVectors(temp_vector, offset);
            points.push(temp_vector);
        }   
    }
    return points;
}


/**
 * Check is a point is inside a cube
 * @param {Array 1d} point 3d vector 
 * @param {Array 1d} p the position of cube
 * @param {Array 1d} s the size of cube
 * @param {Array 1d} r the rotation of cube
 * @param {Float} error the amount of error you are willing to accept (0 for exact check)
 * @returns {Boolean} true if inside else false
 */
function checkPointInsideCube(point, p, s, r, error) {
    let new_point = rotateVector3DInverse(point, r);
    let new_p = rotateVector3DInverse(p, r);
    new_point = addVectors(new_point, scaleVector(new_p, -1));
    if (
        (-s[0]/2 <= (new_point[0] + error)) && ((new_point[0] - error) <= s[0]/2) &&
        (-s[1]/2 <= (new_point[1] + error)) && ((new_point[1] - error) <= s[1]/2) &&
        (-s[2]/2 <= (new_point[2] + error)) && ((new_point[2] - error) <= s[2]/2)
    ) {
        return true;
    } else {
        return false;
    }
}


/**
 * Housekeeping functions
 */


/**
 * Gets the real p vector instead of the fake one
 * @param {Array 1d} p the position of cube (fake)
 * @param {Array 1d} s the size of cube
 * @param {Array 1d} r the rotation of cube
 * @param {Array 1d} p the position of cube
 * @returns {Array 1d} the position of cube (real)
 */
function getRealP(p,s,r) {
    return addVectors(p, rotateVector3D([0, s[1] / 2,0], r));
}


function getRealRProperty(map_json) {
    for (let i = 0; i < map_json.objects.length; i++) {
        if (!('r' in map_json.objects[i])) {
            map_json.objects[i]['r'] = [0,0,0];
        }
    }
    return map_json;
}


/**
 * Gets the real s property
 * @param {JSON} map_json
 * @returns {JSON} map_json
 */
function getRealSProperty(map_json) {
    for (let i = 0; i < map_json.objects.length; i++) {
        if ('si' in map_json.objects[i]) {
            map_json.objects[i]['s'] = [
                map_json.xyz[map_json.objects[i]['si'] * 3 + 0],
                map_json.xyz[map_json.objects[i]['si'] * 3 + 1],
                map_json.xyz[map_json.objects[i]['si'] * 3 + 2]
            ];
            delete map_json.objects[i]['si'];
        }
    }
    return map_json;
}


/**
 * Gets the fake face_index from the real one
 * @param {Integer} face_index (real) between 0 and 5 both inclusive
 * @returns {Integer} the face index (fake) between 0 and 5 both inclusive
 */
function getFakeFaceIndex(face_index) {
    switch(face_index) {
        case 0:
            return 4;
        case 1:
            return 1;
        case 2:
            return 2;
        case 3:
            return 5;
        case 4:
            return 0;
        case 5:
            return 3;
    }
}


/**
 * Remove faces given a map_export
 * @param {JSON} map_export
 * @returns {JSON} map_export
 */
function removeFaces(map_export) 
{
    let map_json = JSON.parse(map_export);
    map_json = getRealRProperty(map_json);
    map_json = getRealSProperty(map_json);    
    let map_objects = map_json.objects;
    let object_count = map_objects.length;
    let face_count_old = 0;
    let face_count_new = 0;
    for(let i = 0; i < object_count ; i++) {
        if ('i' in map_objects[i]) {
            continue;
        }
        console.log('Processing object ' + String(i));
        map_json.objects[i]['h'] = i;
        for(let face = 0; face < 6 ; face++) {
            let face_fake = getFakeFaceIndex(face);
            if ('f' in map_objects[i]) {
                let old_face = parseInt(map_objects[i]['f'], 16);
                if (old_face.toString(2).charAt(face_fake) == "0") {
                    // do nothing, 0 -> already deactivated
                    continue;
                }
            }
            let real_p = getRealP(map_objects[i]['p'], map_objects[i]['s'], map_objects[i]['r']);
            let face_points = getFacePoints(real_p, map_objects[i]['s'], map_objects[i]['r'], face, 10);
            let covered = true;
            for (let j = 0; j < face_points.length; j++) {
                let covered_once = false;
                for(let k = 0; k < object_count ; k++) {
                    if (i == k) {
                        continue;
                    }
                    if ('i' in map_objects[k]) {
                        continue;
                    }
                    let real_p_2 = getRealP(map_objects[k]['p'], map_objects[k]['s'], map_objects[k]['r']);
                    if (checkPointInsideCube(face_points[j], real_p_2, map_objects[k]['s'], map_objects[k]['r'], 0.1)) {
                        covered_once = true;
                        // map_json.objects.push({"p":face_points[j],"s":[0.3,0.3,0.3],"c":"#ff6600","h":k})
                        break;
                    } else {
                        // map_json.objects.push({"p":face_points[j],"s":[0.3,0.3,0.3],"c":"#0066ff"})
                    }
                }
                if (covered_once == false) {
                    covered = false;
                    break;
                }
            }
            if (covered) {
                if ('f' in map_objects[i]) {
                    let old_face = parseInt(map_objects[i]['f'], 16);
                    if (old_face.toString(2).charAt(face_fake) == "0") {
                        // do nothing, 0 -> already deactivated
                    } else {
                        face_count_old += 1;
                        old_face -= Math.pow(2, 5 - face_fake);
                        map_json.objects[i]['f'] = old_face.toString(16);
                    }
                } else {
                    face_count_old += 1;
                    map_json.objects[i]['f'] = (Math.pow(2, 6) - 1 - Math.pow(2, 5 - face_fake)).toString(16);
                }
            } else {
                face_count_old += 1;
                face_count_new += 1;
            }
        }
    }
    console.log("Old face count " + String(face_count_old));
    console.log("New face count " + String(face_count_new));
    console.log("Reduction " + (100*(face_count_old - face_count_new)/ face_count_old).toFixed(2) + "%");
    return map_json;
}


/**
 * The main function
 * @returns 
 */
function main() {
    try {
        let map_export = document.getElementById('input').value;
        map_export = removeFaces(map_export);
        map_export = JSON.stringify(map_export)
        let output = document.getElementById('output');
        output.value = map_export;
    } catch (e) {
        alert("Error: " + e);
    }
}

File: c/face-remove.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#define DIVS 10
#define OBJ_COUNT_MAX 10000
#define MAX_DIGITS 10


typedef struct matrix {
    double cells[3][3];
} Matrix;


typedef struct vector {
    double cells[3];
} Vector;

typedef struct vector_list {
    int length;
    Vector list[DIVS + 1];
} VectorList;

typedef struct cube {
    Vector p;
    Vector s;
    Vector r;
    char f[6];
} Cube;

typedef struct cube_list {
    int length;
    Cube list[OBJ_COUNT_MAX];
} CubeList;


Matrix newMatrix() {
    double cells[3][3] = {{0,0,0},{0,0,0},{0,0,0}};
    Matrix result;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            result.cells[i][j] = cells[i][j];
        }
    }
    return result;
}


Matrix newMatrixValues(double cells[3][3]) {
    Matrix result;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            result.cells[i][j] = cells[i][j];
        }
    }
    return result;
}


Matrix matrixMultiply(Matrix matrix_1, Matrix matrix_2) {
    Matrix result = newMatrix();
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            for (int k = 0; k < 3; k ++) {
                result.cells[i][j] += matrix_1.cells[i][k] * matrix_2.cells[k][j];
            }
        }
    }
    return result;
}


Vector newVector() {
    Vector result;
    double cells[3] = {0,0,0};
    for (int i = 0; i < 3; i++) {
        result.cells[i] = cells[i];
    }
    return result;
}


Vector newVectorValues(double cells[3]) {
    Vector result;
    for (int i = 0; i < 3; i++) {
        result.cells[i] = cells[i];
    }
    return result;
}


VectorList newVectorList() {
    VectorList vl;
    vl.length = 0;
    return vl;
}


Vector scaleVector(Vector vector, double scale) {
    Vector result;
    for (int i = 0; i < 3; i++) {
        result.cells[i] = vector.cells[i] * scale;
    }
    return result;
}

Vector addVectors(Vector vector_1, Vector vector_2) {
    Vector result = newVector();
    for (int i = 0; i < 3; i++) {
        result.cells[i] = vector_1.cells[i] + vector_2.cells[i];
    }
    return result;
}

Vector matrixVectorMultiply(Matrix matrix_1, Vector matrix_2) {
    Vector result = newVector();
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            result.cells[i] += matrix_1.cells[i][j] * matrix_2.cells[j];
        }
    }
    return result;
}


Vector rotateVector(Vector vector, Vector r) {
    Vector result = newVector();
    for (int i = 0; i < 3; i++) {
        result.cells[i] = vector.cells[i];
    }

    double rX[3][3] = {
        {1, 0, 0},
        {0, cos(r.cells[0]), -sin(r.cells[0])},
        {0, sin(r.cells[0]), cos(r.cells[0])}
    };

    double rY[3][3] = {
        {cos(r.cells[1]), 0, sin(r.cells[1])},
        {0, 1, 0},
        {-sin(r.cells[1]), 0, cos(r.cells[1])}
    };

    double rZ[3][3] = {
        {cos(r.cells[2]), -sin(r.cells[2]), 0},
        {sin(r.cells[2]), cos(r.cells[2]), 0},
        {0, 0, 1}
    };

    Matrix temp_1 = matrixMultiply(newMatrixValues(rX), newMatrixValues(rY));
    Matrix temp_2 = matrixMultiply(temp_1, newMatrixValues(rZ));
    result = matrixVectorMultiply(temp_2, result);

    return result;
}


Vector rotateVectorInverse(Vector vector, Vector r) {
    Vector result = newVector();
    for (int i = 0; i < 3; i++) {
        result.cells[i] = vector.cells[i];
    }

    double rX[3][3] = {
        {1, 0, 0},
        {0, cos(r.cells[0]), sin(r.cells[0])},
        {0, -sin(r.cells[0]), cos(r.cells[0])}
    };

    double rY[3][3] = {
        {cos(r.cells[1]), 0, -sin(r.cells[1])},
        {0, 1, 0},
        {sin(r.cells[1]), 0, cos(r.cells[1])}
    };

    double rZ[3][3] = {
        {cos(r.cells[2]), sin(r.cells[2]), 0},
        {-sin(r.cells[2]), cos(r.cells[2]), 0},
        {0, 0, 1}
    };

    Matrix temp_1 = matrixMultiply(newMatrixValues(rZ), newMatrixValues(rY));
    Matrix temp_2 = matrixMultiply(temp_1, newMatrixValues(rX));
    result = matrixVectorMultiply(temp_2, result);
    
    return result;
}


Cube newCube() {
    Cube c;
    strcpy(c.f, "");
    c.p = newVector();
    c.s = newVector();
    c.r = newVector();
    return c;
}


CubeList newCubeList() {
    CubeList cl;
    cl.length = 0;
    return cl;
}


VectorList getSideVectors(Cube c) {
    double s1[3] = {c.s.cells[0], 0, 0};
    Vector temp_1 = rotateVector(newVectorValues(s1), c.r);

    double s2[3] = {0, c.s.cells[0], 0};
    Vector temp_2 = rotateVector(newVectorValues(s2), c.r);

    double s3[3] = {0, 0, c.s.cells[2]};
    Vector v3 = newVectorValues(s3);
    Vector temp_3 = rotateVector(v3, c.r);

    VectorList result;
    result.list[0] = temp_1;
    result.list[1] = temp_2;
    result.list[2] = temp_3;
    result.length = 3;

    return result;
}


VectorList getFacePoints(Cube c, int face) {
    VectorList side_vectors = getSideVectors(c);

    Vector vector_1 = side_vectors.list[face % 3];
    Vector vector_2 = side_vectors.list[(face + 1) % 3];
    Vector vector_3 = side_vectors.list[(face + 2) % 3];
    int sign = (face % 2) ? -1 : 1;

    Vector offset = addVectors(scaleVector(vector_1, -0.5), scaleVector(vector_2, -0.5));
    offset = addVectors(offset, scaleVector(vector_3, 0.5 * sign));
    offset = addVectors(offset, c.p);
    Vector vector_1_div = scaleVector(vector_1, 1.0 / DIVS);
    Vector vector_2_div = scaleVector(vector_2, 1.0 / DIVS);

    VectorList points;
    for (int i = 0; i <= DIVS; i ++) {
        for (int j = 0; j <= DIVS; j ++) {
            Vector temp_vector = addVectors(scaleVector(vector_1_div, i), scaleVector(vector_2_div, j));
            temp_vector = addVectors(temp_vector, offset);
            points.list[i * DIVS + j] = temp_vector;
        }   
    }
    return points;
}


int checkPointInsideCube(Vector v, Cube c, double error) {
    Vector new_point = rotateVectorInverse(v, c.r);
    Vector new_p = rotateVectorInverse(c.p, c.r);
    new_point = addVectors(new_point, scaleVector(new_p, -1));
    if (
        (-c.s.cells[0]/2 <= (new_point.cells[0] + error)) && ((new_point.cells[0] - error) <= c.s.cells[0]/2) &&
        (-c.s.cells[1]/2 <= (new_point.cells[1] + error)) && ((new_point.cells[1] - error) <= c.s.cells[1]/2) &&
        (-c.s.cells[2]/2 <= (new_point.cells[2] + error)) && ((new_point.cells[2] - error) <= c.s.cells[2]/2)
    ) {
        return 1;
    } else {
        return 0;
    }
}


Cube getRealP(Cube c) {
    double s2[3] = {0, c.s.cells[1] / 2,0};
    c.p = addVectors(c.p, rotateVector(newVectorValues(s2), c.r));
    return c;
}


int getFakeFaceIndex(int face_index) {
    switch(face_index) {
        case 0:
            return 4;
        case 1:
            return 1;
        case 2:
            return 2;
        case 3:
            return 5;
        case 4:
            return 0;
        case 5:
            return 3;
        default:
            return 0;
    }
}


CubeList readInCubes() {
    CubeList cl;
    cl.length = 0;

    char *ptr;
    char t[MAX_DIGITS];
    while(1) {
        Cube c;

        if (scanf("%s", c.f) == EOF) {
            break;
        }

        for (int i = 0; i < 3; i++) {
            scanf("%s", t);
            c.p.cells[i] = strtod(t, &ptr);
        }

        for (int i = 0; i < 3; i++) {
            scanf("%s", t);
            c.s.cells[i] = strtod(t, &ptr);
        }

        for (int i = 0; i < 3; i++) {
            scanf("%s", t);
            c.r.cells[i] = strtod(t, &ptr);
        }

        cl.list[cl.length] = c;
        cl.length += 1;
    }

    return cl;
}


void printOutCubes(CubeList cl) {
    for (int i = 0; i < cl.length; i++) {
        printf("%s\n", cl.list[i].f);

        // for (int j = 0; j < 3; j++) {
        //     printf("%f\n", cl.list[i].p.cells[j]);
        // }

        // for (int j = 0; j < 3; j++) {
        //     printf("%f\n", cl.list[i].s.cells[j]);
        // }

        // for (int j = 0; j < 3; j++) {
        //     printf("%f\n", cl.list[i].r.cells[j]);
        // }
    }
}


int main(int argc, char const *argv[]) {
    // double s[3] = {1, 1, 1};
    // Vector v = newVectorValues(s);
    // double r[3] = {0, 1, 0};
    // Vector w = newVectorValues(r);
    // Vector y = rotateVectorInverse(v, w);
    // printf("%f, %f, %f \n", y.cells[0], y.cells[1], y.cells[2]);

    CubeList cl = readInCubes();
    printOutCubes(cl);

    return 0;
}
